### PR TITLE
S3 downloading/uploading

### LINKED ISSUE #309: encrypt shieldhit binary

### LINKED ISSUE #310: write a script which gets binary from S3 and decrypts it

### PR DESCRIPTION
**Pull request by lequ0n**

### CODE REVIEW

**Comment by lequ0n** [comment]
In current state four different variables separated by spaces are stored in single environmental variable `endpoint, access_key, secret_key, encryption_key = os.environ["YAPTIDE_S3_CONFIG"].split()`
Endpoint, access_key and secret_key are used to connect with s3. Encryption_key is used in encryption.
I wasn't sure where to place `s3_uploader.py` and currently it is in `/yaptide/utils`.

File: yaptide/utils/s3_uploader.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def encrypt_file(file_path: Path, key: str) -> Path:
+    """Encrypts a file using Fernet"""
+    if not file_path.is_file():
+        raise ValueError("Invalid file path or file does not exist.")
+    with open(file_path, "rb") as file:
+        original = file.read()
+    fernet = Fernet(key)
+    encrypted = fernet.encrypt(original)
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: I'd rather create a function which encrypts files and gives back on return a string with file contents. Then file saving and deleting can be easily handled by callers

File: yaptide/utils/s3_uploader.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    try:
+        # Upload encrypted file to S3 bucket
+        print("Uploading file.")
+        s3_client.upload_file(encrypted_file_path, bucket, os.path.basename(file_path))
+        print("Upload successful.")
+    except ClientError as e:
+        raise Exception("Upload failed:", e.response["Error"]["Message"]) from e
+
+
+def main():
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: Use click library as in simulators.py to handle command line options

File: yaptide/utils/s3_uploader.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    # Encrypt file
+    encrypted_file_path = encrypt_file(file_path, encryption_key)
+    try:
+        # Upload encrypted file to S3 bucket
+        print("Uploading file.")
+        s3_client.upload_file(encrypted_file_path, bucket, os.path.basename(file_path))
+        print("Upload successful.")
+    except ClientError as e:
+        raise Exception("Upload failed:", e.response["Error"]["Message"]) from e
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: Replace encrypted_file_path with encrypted_file_contents. Use `with tempfile.NamedTemporaryFile() as temp_file:` to create temporary file to store encrypted shieldhit binary which will be afterwards deleted

File: yaptide/utils/s3_uploader.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def main():
+    """Responsible for checking input arguments and calling upload_file_to_s3"""
+    if "YAPTIDE_S3_CONFIG" not in os.environ:
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: Move "YAPTIDE_S3_CONFIG" to a module level variable.

File: yaptide/utils/s3_uploader.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        raise Exception("YAPTIDE_S3_CONFIG environment variable not set.")
+    endpoint, access_key, secret_key, encryption_key = os.environ["YAPTIDE_S3_CONFIG"].split()
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: I would create four separate variables for each of items. All of them should be present to proceed with upload

File: yaptide/utils/s3_uploader.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    if len(sys.argv) != 3:
+        raise Exception("Wrong number of arguments <path> <bucket> needed.")
+    path = Path(sys.argv[1])
+    bucket = sys.argv[2]
+    upload_file_to_s3(endpoint, access_key, secret_key, encryption_key, path, bucket)
```
Comment (grzanka) [inline yaptide/utils/s3_uploader.py]: Command line script could use `dotenv` library to read the enviromental variables or .env file. It should also accept command line options which overwrite the .env/ env variables. See https://click.palletsprojects.com/en/8.1.x/options/#values-from-environment-variables

**Review by grzanka — CHANGES_REQUESTED** [review]

File: yaptide/admin/simulators.py [hunk 7]
```diff
@@ -115,5 +221,34 @@ def install(**kwargs):
...
+        'S3 endpoint is required specify with --endpoint',
+        'S3 access key not found in environment variables specify with --access_key',
+        'S3 secret key not found in environment variables specify with --secret_key',
+        'S3 encryption key not found in environment variables specify with --encryption_key'
+    ]
+    for arg, message in zip(args, messages):
+        if not kwargs[arg] or kwargs[arg] is None:
+            click.echo(message)
+            return
+    if not os.path.isfile(kwargs['file']):
```
Comment (grzanka) [inline yaptide/admin/simulators.py]: Move to pathlib

File: yaptide/admin/simulators.py [hunk 8]
```diff
@@ -70,35 +79,134 @@ def install_simulator(name: SimulatorType) -> bool:
...
-            click.echo(f"Extracting {temp_file_archive} to {installation_path}")
-            if temp_file_archive.suffix == '.gz':
-                extract_shieldhit_from_tar_gz(temp_file_archive, Path(tmpdir_name), 'shieldhit')
-            elif temp_file_archive.suffix == '.zip':
-                extract_shieldhit_from_zip(temp_file_archive, Path(tmpdir_name), 'shieldhit.exe')
+        shieldhit_installed = False
+        if all([endpoint, access_key, secret_key, encryption_key]):
+            click.echo('Downloading from S3 bucket')
+            shieldhit_installed = download_shieldhit_from_s3()
+        if not shieldhit_installed:
```
Comment (grzanka) [inline yaptide/admin/simulators.py]:
```suggestion
            shieldhit_downloaded_from_s3 = download_shieldhit_from_s3()
        if not shieldhit_downloaded_from_s3:
```

File: yaptide/admin/simulators.py [hunk 9]
```diff
@@ -70,35 +79,134 @@ def install_simulator(name: SimulatorType) -> bool:
...
+        "s3",
+        aws_access_key_id=access_key,
+        aws_secret_access_key=secret_key,
+        endpoint_url=endpoint
+    )
+    destination_file_path = installation_path / "shieldhit"
+    with tempfile.TemporaryFile() as temp_file:
+        # Download file from s3 bucket
+        try:
+            s3_client.download_fileobj("shieldhit", "shieldhit", temp_file)
```
Comment (grzanka) [inline yaptide/admin/simulators.py]: Extract bucket_name and key_name to parameters in the `download_shieldhit_from_s3` method with reasonable default values

File: yaptide/admin/simulators.py [hunk 10]
```diff
@@ -70,35 +79,134 @@ def install_simulator(name: SimulatorType) -> bool:
...
+        temp_file.seek(0)
+        encrypted_data = temp_file.read()
+    # Decrypt downloaded file
+    click.echo("Decrypting downloaded file")
+    fernet = Fernet(encryption_key)
+    decrypted_data = fernet.decrypt(encrypted_data)
+    with open(destination_file_path, "wb") as f:
+        f.write(decrypted_data)
+    # Permission to execute
+    os.chmod(destination_file_path, 0o700)
```
Comment (grzanka) [inline yaptide/admin/simulators.py]:
```suggestion
    # Decrypt downloaded file
    click.echo("Decrypting downloaded file")
    fernet = Fernet(encryption_key)
    decrypted_data = fernet.decrypt(encrypted_data)
    with open(destination_file_path, "wb") as f:
        f.write(decrypted_data)
    # Permission to execute
    os.chmod(destination_file_path, 0o700)
```

I would move that section entirely inside `with tempfile.TemporaryFile() as temp_file:` construct

File: yaptide/admin/simulators.py (hunk 9 above)
Comment (grzanka) [reply yaptide/admin/simulators.py]: Also use named parameters

File: yaptide/admin/simulators.py [hunk 11]
```diff
@@ -70,35 +79,134 @@ def install_simulator(name: SimulatorType) -> bool:
...
+        temp_file.seek(0)
+        encrypted_data = temp_file.read()
+        # Decrypt downloaded file
+        click.echo("Decrypting downloaded file")
+        fernet = Fernet(encryption_key)
+        decrypted_data = fernet.decrypt(encrypted_data)
+        with open(destination_file_path, "wb") as f:
+            f.write(decrypted_data)
+        # Permission to execute
+        os.chmod(destination_file_path, 0o700)
```
Comment (grzanka) [inline yaptide/admin/simulators.py]: Use pathlib

Comment (grzanka) [reply yaptide/admin/simulators.py]: I think you already have a path like object

File: requirements.txt [hunk 12]
```diff
@@ -14,3 +14,5 @@ requests==2.29.0
 scipy==1.10.1  # needed by converter submodule
 SQLAlchemy==1.4.44
 Werkzeug==2.3.4
+boto3==1.26.142
+cryptography==40.0.2
```
Comment (grzanka) [inline requirements.txt]: Add newline at the end

File: tests/test_secrets.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import os
+
+def test_setting_secrets():
+    """Check if secrets are set"""
+    assert os.environ.get('S3_ENDPOINT') is not None
```
Comment (grzanka) [inline tests/test_secrets.py]:
```suggestion
    assert 'S3_ENDPOINT' in os.environ
```

**Comment by grzanka** [comment]
@lequ0n I think its my task to upload the shielhit binary and put proper secrets in the Github repo, I will do it today.

File: yaptide/admin/simulators.py [hunk 14]
```diff
@@ -23,6 +27,11 @@ def run():
     """Manage simulators"""
 
 
+load_dotenv()
+endpoint = os.getenv('S3_ENDPOINT')
+access_key = os.getenv('S3_ACCESS_KEY')
+secret_key = os.getenv('S3_SECRET_KEY')
+encryption_key = os.getenv('S3_ENCRYPTION_KEY')
```
Comment (grzanka) [inline yaptide/admin/simulators.py]: I would remove these module level variables and make them a parameters of all respective functions.

Take a look at this not working use case:
```
python yaptide/admin/simulators.py upload --bucket shieldhit-demo --file ./1/shield_hit12a_x86_64_demo_gfortran_v1.0.1/bin/shieldhit --endpoint https://s3.cloud.cyfronet.pl --access_key XXX --secret_key XXX --encryption_key XXXX
Uploading 1/shield_hit12a_x86_64_demo_gfortran_v1.0.1/bin/shieldhit to S3 bucket shieldhit-demo
Endpoint: None
Access key: None
No credentials found. Check your access key and secret key. Unable to locate credentials
```

**Comment by grzanka** [comment]
Setting in `.env` following line:
```
S3_ENCRYPTION_KEY = 'XXXX'
```

gives me:
```
Traceback (most recent call last):
  File "yaptide/admin/simulators.py", line 263, in <module>
    run()
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "yaptide/admin/simulators.py", line 258, in upload
    if upload_file_to_s3(kwargs['bucket'], Path(kwargs['file'])):
  File "yaptide/admin/simulators.py", line 186, in upload_file_to_s3
    encrypted_file_contents = encrypt_file(file_path)
  File "yaptide/admin/simulators.py", line 202, in encrypt_file
    fernet = Fernet(encryption_key)
  File "/data/grzanka/workspace/yaptide/venv/lib/python3.8/site-packages/cryptography/fernet.py", line 39, in __init__
    raise ValueError(
ValueError: Fernet key must be 32 url-safe base64-encoded bytes.
```

**Comment by grzanka** [comment]
Maybe Fernet with password be a better option, see
https://cryptography.io/en/latest/fernet/#using-passwords-with-fernet

**Comment by grzanka** [comment]
Fix tests, some bucket names seems hardcoded to wrong values:

`>       assert download_shieldhit_from_s3(bucket="shieldhit", key="shieldhit", installation_path=installation_path) is True`

Bucket and key should be extracted from env variables

**Comment by grzanka** [comment]
image
image

You should now be able to make some tests, the secrets are on Github Actions

File: yaptide/admin/simulators.py [hunk 15]
```diff
@@ -115,5 +227,34 @@ def install(**kwargs):
         click.echo(f'Simulator {sim_type.name} installed')
 
 
+@run.command
+@click.option('--bucket', help='S3 bucket name')
+@click.option('--file', help='Path to file to upload')
+@click.option('--endpoint', envvar='S3_ENDPOINT', default=endpoint, help='S3 endpoint')
+@click.option('--access_key', envvar='S3_ACCESS_KEY', default=access_key, help='S3 access key')
+@click.option('--secret_key', envvar='S3_SECRET_KEY', default=secret_key, help='S3 secret key')
+@click.option('--encryption_key', envvar='S3_ENCRYPTION_KEY', default=encryption_key, help='S3 encryption key')
```
Comment (grzanka) [inline yaptide/admin/simulators.py]:
```suggestion
@click.option('--encryption_key', envvar='S3_ENCRYPTION_KEY', default=encryption_key, help='binary encryption key')
```

Fernet has strict rules on how the key should look like (i.e. string that is converted to extactly 32 bytes). Either provide a clear instruction how to generate such key (random selection of letters doesn't work) or swithc to a different approach, i.e. using passwords in Fernet

**Comment by lequ0n** [comment]
Right now encryption_key is derived from password and salt. We need additional secrets `S3_ENCRYPTION_PASSWORD`, `S3_ENCRYPTION_SALT` (`S3_ENCRYPTION_KEY` is no longer needed). Also we need more repository variables for testing on both linux and windows. Bucket `S3_SHIELDHIT_BUCKET` may be same for both  but we need separate keys `S3_LINUX_SHIELDHIT_KEY` and `S3_WINDOWS_SHIELDHIT_KEY` for downloading appropriate files (in current yml there are also separate buckets).

**Comment by grzanka** [comment]
> Right now encryption_key is derived from password and salt. We need additional secrets `S3_ENCRYPTION_PASSWORD`, `S3_ENCRYPTION_SALT` (`S3_ENCRYPTION_KEY` is no longer needed). Also we need more repository variables for testing on both linux and windows. Bucket `S3_SHIELDHIT_BUCKET` may be same for both but we need separate keys `S3_LINUX_SHIELDHIT_KEY` and `S3_WINDOWS_SHIELDHIT_KEY` for downloading appropriate files (in current yml there are also separate buckets).

I will upload necessary keys today.

**Comment by grzanka** [comment]
At a later stage we could think of using tags and versions in S3 to avoid too many secrets, but this can go in a subsequent PR

File: .github/workflows/test-run.yml [hunk 16]
```diff
@@ -7,6 +7,12 @@ on:
     branches: [master]
 
   workflow_dispatch:
+env:
```
Comment (grzanka) [inline .github/workflows/test-run.yml]: Use secrets only where needed. Making then per-job global variable exposes some security risk, move assignment of secrets to env variables only to the steps which needs them

Comment (grzanka) [reply .github/workflows/test-run.yml]: Include secrets in coverage job in this file as well

File: yaptide/admin/simulators.py [hunk 17]
```diff
@@ -70,35 +85,152 @@ def install_simulator(name: SimulatorType) -> bool:
...
+    # Check if bucket exists and create if not
+    if bucket not in [bucket["Name"] for bucket in s3_client.list_buckets()["Buckets"]]:
+        click.echo("Bucket does not exist. Creating bucket.")
+        s3_client.create_bucket(Bucket=bucket)
+
+    # Encrypt file
+    encrypted_file_contents = encrypt_file(file_path, encryption_password, encryption_salt)
+    try:
+        # Upload encrypted file to S3 bucket
+        click.echo("Uploading file.")
```
Comment (grzanka) [inline yaptide/admin/simulators.py]: Add filename in the click echo message

**Comment by grzanka** [comment]
Before merging I need to check if procedure with https://github.com/yaptide/deploy still works

**Review by grzanka — APPROVED** [review]

**Pull request merged**